aesd-char-driver/main.c

```diff
@@ -0,0 +1,398 @@
+     * 0 < return < count, only portion of bytes transferred
+     * return == 0, end of file, no data read into buf
+     * return < 0, error occurred
+     */
+    
+    size_t read_entry_offset_rtn = 0;
+    size_t new_count = 0;
+    struct aesd_dev *dev = (struct aesd_dev*)filp->private_data;
+    struct aesd_buffer_entry *read_entry;
+    ssize_t retval = 0;
+
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+    
+/* obtain mutex, exit on failure */
+    if(mutex_lock_interruptible(&dev->mutex) != 0)
+    {
+        retval = -ERESTARTSYS;
+        goto exit;
+    }
+
+/* parameter error handling */
+    if(filp == NULL || buf == NULL || f_pos == NULL)
+    {
+        retval = -EFAULT;
+        goto exit;
+    }
+/* attempting to read 0 bytes, always return 0 */
+    if(count == 0)
+    {
+        retval = 0;
+        goto exit;
+    }
+
+/* end goal is to find the entry and copy to user space */
+    read_entry = aesd_circular_buffer_find_entry_offset_for_fpos(&dev->buffer, *f_pos, &read_entry_offset_rtn);
+    if(read_entry == NULL)
+        goto exit;
+  
+/* check if count would extend past last byte of read entry when starting at desired offset */
+    if((read_entry->size - read_entry_offset_rtn) < count)
+        new_count = read_entry->size - read_entry_offset_rtn;
+    else
+        new_count = count;
+    *f_pos += new_count;
+
+/* returns 0 on success, >0 is number of bytes not read */
+    if(copy_to_user(buf, (read_entry->buffptr + read_entry_offset_rtn), new_count))
+    {
+        retval = -EFAULT;
+        goto exit;
+    }
+
+/* return total bytes read */
+    retval = new_count;
+
+exit:
+    mutex_unlock(&dev->mutex);
+
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    /**
+     * TODO: handle write
+     * 
+     * all write appends to command being written until newline
+     * write to the command buffer when newline is received
+     * 
+     * return == count, success
+     * 0 < return < count, partial write, retry command
+     * return == 0, nothing written, retry command
+     * return < 0, error occurred
+     * 
+     */
+
+    size_t i = 0;
+    bool cmd_end_flag = false;
+    size_t cmd_len = 0;
+    char *input_buffer = NULL;
+    const char *overwrite = NULL;
+    struct aesd_dev *dev = filp->private_data;  /* get pointer to our char device */
+    ssize_t retval = 0;
+
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+
+/* parameter error handling */
+    if(filp == NULL || buf == NULL || f_pos == NULL)
+    {
+        retval = -EFAULT;
+        goto exit;
+    }
+
+/* obtain mutex, exit on failure */
+    if(mutex_lock_interruptible(&dev->mutex) != 0)
+    {
+        retval = -ERESTARTSYS;
+        goto exit;
+    }
+
+/* allocate input buffer to store the entire write */
+    input_buffer = (char *)kmalloc(count, GFP_KERNEL);
```

Comment:
@mash9756 does this line through line 202 need to be in the locking interval?  What is `dev->mutex` locking?  See discussion in class about this today.